tensorflow/core/platform/s3/s3_file_system.cc

```diff
@@ -583,6 +667,234 @@ Status S3FileSystem::GetFileSize(const string& fname, uint64* file_size) {
   return Status::OK();
 }
 
+void S3FileSystem::MultiPartCopyCallback(const Aws::S3::Model::UploadPartCopyRequest& request,
+                                         const Aws::S3::Model::UploadPartCopyOutcome& uploadPartCopyOutcome,
+                                         const std::shared_ptr<const Aws::Client::AsyncCallerContext>& context) {
+
+  std::shared_ptr<tensorflow::MultiPartCopyAsyncContext> multiPartContext =
+  std::const_pointer_cast<tensorflow::MultiPartCopyAsyncContext>(std::static_pointer_cast<const tensorflow::MultiPartCopyAsyncContext>(context));
+
+  {
+    std::unique_lock <std::mutex> lock(*multiPartContext->multi_part_copy_mutex);
+
+    Status status;
+    if (uploadPartCopyOutcome.IsSuccess()) {
+      // success
+      Aws::String eTag = uploadPartCopyOutcome.GetResult().GetCopyPartResult().GetETag();
+      multiPartContext->eTag = eTag;
+      status = Status::OK();
+    } else {
+      LOG(ERROR) << "Error when copying part " << multiPartContext->partNumber << " " << uploadPartCopyOutcome.GetError().GetMessage();
+      status = errors::Unknown(uploadPartCopyOutcome.GetError().GetExceptionName(), ": ",
+                               uploadPartCopyOutcome.GetError().GetMessage());
+    }
+
+    (*multiPartContext->finishedPartStates)[multiPartContext->partNumber] = multiPartContext->incompletePartStates->at(multiPartContext->partNumber);
+    multiPartContext->finishedPartStates->at(multiPartContext->partNumber).status = status;
+    multiPartContext->incompletePartStates->erase(multiPartContext->partNumber);
+    // Notify the thread that started the operation
+    multiPartContext->multi_part_copy_cv->notify_one();
+  }
+
+}
+
+Status S3FileSystem::CopyFile(const Aws::String& source_bucket, const Aws::String& source_key,
+                              const Aws::String& target_bucket, const Aws::String& target_key) {
+  Aws::String source = Aws::String((source_bucket + "/" + source_key).c_str());
+  Aws::String source_full_path = Aws::String("s3://") + source;
+  uint64 file_length;
+  TF_RETURN_IF_ERROR(this->GetFileSize(std::string(source_full_path.c_str()), &file_length));
+  int num_parts;
+  if (file_length <= multi_part_copy_part_size_) {
+    num_parts = 1;
+  } else {
+    num_parts = ceil((float) file_length / multi_part_copy_part_size_);
+  }
+  
+  if (num_parts == 1) {
+    return SimpleCopy(source, target_bucket, target_key);
+  } else if (num_parts > 10000) {
+    std::ostringstream s;
```

Comment:
Let's use `tensorflow::strings::StrCat` instead of `std::ostringstream`

Reply:
Nice function, handles lot of types of arguments :) 
Done